code/modules/food_and_drinks/drinks/drinks/cans.dm

```diff
@@ -49,15 +86,47 @@
 /obj/item/reagent_containers/food/drinks/cans/afterattack(obj/target, mob/user, proximity)
 	if(!proximity)
 		return
-	if(istype(target, /obj/structure/reagent_dispensers) && (canopened == 0))
+	if(istype(target, /obj/structure/reagent_dispensers) && (canopened == FALSE))
 		to_chat(user, "<span class='notice'>You need to open the drink!</span>")
 		return
-	else if(target.is_open_container() && (canopened == 0))
+	else if(target.is_open_container() && (!canopened))
 		to_chat(user, "<span class='notice'>You need to open the drink!</span>")
 		return
 	else
 		return ..(target, user, proximity)
 
+/obj/item/reagent_containers/food/drinks/cans/after_throw(mob/user)
+	if(shaken < 5)
+		shaken++
+	else
+		burst(user)
+
+/obj/item/reagent_containers/food/drinks/cans/proc/foam_up()
+	if(reagents.total_volume)
+		var/datum/effect_system/foam_spread/sodafizz = new
+		sodafizz.set_up(1, get_turf(src), reagents)
+		sodafizz.start()
+
+
+/obj/item/reagent_containers/food/drinks/cans/proc/burst(mob/user)
+	if(!bursting)
+		addtimer(CALLBACK(src, .proc/resetburst), 10 SECONDS)
+		bursting = TRUE
+		burstchance = 1
+		return
+	if(burstchance < 10)
+		burstchance++
+	if(prob((burstchance * 5)))
+		attack_self(user, TRUE)
```

Comment:
Please don't call `attack_self` from code if a user doesn't attack the object itself. It makes the code very impure and is prone for errors.
It's best to make a new proc for the bursting logic and then call that from `attack_self` and in `burst` instead